src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,846 @@
+                    if (CompletionHelper.TryCreateMatchResult(_completionHelper, roslynItem, item, _filterText,
+                        _initialRoslynTriggerKind, FilterReason, _recentItemsManager.RecentItems, _highlightMatchingPortions, currentIndex,
+                        out var matchResult))
+                    {
+                        list.Add(matchResult);
+                        currentIndex++;
+                    }
+                }
+
+                // Sort the items by pattern matching results.
+                // Note that we want to preserve the original alphabetical order for items with same pattern match score,
+                // but `List<T>.Sort` isn't stable. Therefore we have to add a monotonically increasing integer
+                // to `MatchResult` to achieve this.
+                list.Sort(MatchResult<VSCompletionItem>.SortingComparer);
+
+                itemsToBeIncluded = list;
+                return disposer;
+
+            }
+
+            private ItemSelection? HandleNormalFiltering(IReadOnlyList<MatchResult<VSCompletionItem>> items)
+            {
+                // Not deletion.  Defer to the language to decide which item it thinks best
+                // matches the text typed so far.
+
+                // Ask the language to determine which of the *matched* items it wants to select.
+                var matchingItems = items.Where(r => r.MatchedFilterText)
+                                               .SelectAsArray(t => (t.RoslynCompletionItem, t.PatternMatch));
+
+                var chosenItems = FilterMethod(matchingItems, _filterText);
+
+                int selectedItemIndex;
+                VSCompletionItem? uniqueItem = null;
+                MatchResult<VSCompletionItem> bestOrFirstMatchResult;
+
+                if (chosenItems.Length == 0)
+                {
+                    // We do not have matches: pick the one with longest common prefix or the first item from the list.
+                    selectedItemIndex = 0;
+                    bestOrFirstMatchResult = items[0];
+
+                    var longestCommonPrefixLength = bestOrFirstMatchResult.RoslynCompletionItem.FilterText.GetCaseInsensitivePrefixLength(_filterText);
+
+                    for (var i = 1; i < items.Count; ++i)
+                    {
+                        var item = items[i];
+                        var commonPrefixLength = item.RoslynCompletionItem.FilterText.GetCaseInsensitivePrefixLength(_filterText);
+
+                        if (commonPrefixLength > longestCommonPrefixLength)
+                        {
+                            selectedItemIndex = i;
+                            bestOrFirstMatchResult = item;
+                            longestCommonPrefixLength = commonPrefixLength;
+                        }
+                    }
+                }
+                else
+                {
+                    var recentItems = _recentItemsManager.RecentItems;
```

Comment:
this seems to be retrieved just top pass into helper.  Can the helper just retrieve this instead?